`tests.test_pack.TestPackDeltas.test_apply_delta_invalid_opcode` fails due to hitting an assertion

When I'm trying to run tests of 0.23.0 release on Gentoo, I'm seeing the following test failure:

```pytb
======================================================================
FAIL: test_apply_delta_invalid_opcode (tests.test_pack.TestPackDeltas.test_apply_delta_invalid_opcode)
Test apply_delta with an invalid opcode.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/portage/dev-python/dulwich-0.23.0/work/dulwich-0.23.0/tests/test_pack.py", line 232, in test_apply_delta_invalid_opcode
    self.assertRaises(ApplyDeltaError, apply_delta, base, invalid_delta)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 255 vs 9

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (failures=1)
```

This error was quite unclear to me, so I've attempted to run it via pytest instead, and got a better backtrace:

```pytb
________________ TestPackDeltas.test_apply_delta_invalid_opcode ________________

self = <tests.test_pack.TestPackDeltas testMethod=test_apply_delta_invalid_opcode>

    def test_apply_delta_invalid_opcode(self) -> None:
        """Test apply_delta with an invalid opcode."""
        # Create a delta with an invalid opcode (0xff is not valid)
        invalid_delta = [b"\xff\x01\x02"]
        base = b"test base"
    
        # Should raise ApplyDeltaError
>       self.assertRaises(ApplyDeltaError, apply_delta, base, invalid_delta)

tests/test_pack.py:232: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def apply_delta(src_buf, delta):
        """Based on the similar function in git's patch-delta.c.
    
        Args:
          src_buf: Source buffer
          delta: Delta instructions
        """
        if not isinstance(src_buf, bytes):
            src_buf = b"".join(src_buf)
        if not isinstance(delta, bytes):
            delta = b"".join(delta)
        out = []
        index = 0
        delta_length = len(delta)
    
        def get_delta_header_size(delta, index):
            size = 0
            i = 0
            while delta:
                cmd = ord(delta[index : index + 1])
                index += 1
                size |= (cmd & ~0x80) << i
                i += 7
                if not cmd & 0x80:
                    break
            return size, index
    
        src_size, index = get_delta_header_size(delta, index)
        dest_size, index = get_delta_header_size(delta, index)
>       assert src_size == len(src_buf), f"{src_size} vs {len(src_buf)}"
               ^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: 255 vs 9

../dulwich-0.23.0-python3_13/install/usr/lib/python3.13/site-packages/dulwich/pack.py:2519: AssertionError
```

It looks like the tested code has a failing assertion that prevents it from reaching the point where `ApplyDeltaError` would be raised.

I don't know why CI doesn't trigger it — my educated guess is that it's running on optimized code and assertions are stripped.

--- a/NEWS
+++ b/NEWS
@@ -13,6 +13,12 @@
  * Add annotate support as well as ``annotate`` and ``blame``
    commands.  (#245, Jelmer Vernooĳ)
 
+ * Fix ``apply_delta`` to raise ``ApplyDeltaError`` instead of ``AssertionError``
+   when the source buffer size doesn't match the delta header. This issue only
+   affected the pure Python implementation when the Rust extension was not
+   available. The Rust implementation already raised the correct exception.
+   (#1606, Jelmer Vernooĳ)
+
 0.23.0	2025-06-21
 
  * Add basic ``rebase`` subcommand. (Jelmer Vernooĳ)

--- a/dulwich/pack.py
+++ b/dulwich/pack.py
@@ -2516,7 +2516,10 @@ def apply_delta(src_buf, delta):
 
     src_size, index = get_delta_header_size(delta, index)
     dest_size, index = get_delta_header_size(delta, index)
-    assert src_size == len(src_buf), f"{src_size} vs {len(src_buf)}"
+    if src_size != len(src_buf):
+        raise ApplyDeltaError(
+            f"Unexpected source buffer size: {src_size} vs {len(src_buf)}"
+        )
     while index < delta_length:
         cmd = ord(delta[index : index + 1])
         index += 1

